### PR TITLE
Improve ftp session handling

### PR DESCRIPTION
**Pull request by mvangoor**

### CODE REVIEW

File: src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java [hunk 1]
```diff
@@ -332,19 +332,18 @@ public void run() {
             _thread.setName("FtpConn thread " + _thread.getId() + " from " + getClientAddress().getHostAddress());
         }
 
-        _pool = new ThreadPoolExecutor(1, Integer.MAX_VALUE, 60L, TimeUnit.SECONDS, new SynchronousQueue<>(), new CommandThreadFactory(_thread.getName()));
+        _pool = new ThreadPoolExecutor(1, 1, 60L, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), new CommandThreadFactory(_thread.getName()));
```
Comment (k-2-r) [inline src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java:335]: Why do you need a pool executor anymore?
If you want a sequential execution of every commands, you should just remove it.

File: src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java [hunk 2]
```diff
@@ -454,13 +453,18 @@ public void run() {
                 }
 
                 // execute command
-                _pool.execute(new CommandThread(_request, this));
-                if (_request.getCommand().equalsIgnoreCase("AUTH")) {
-                    while (!_securityDataExchangeCompleted && !_stopRequest) {
-                        Thread.sleep(100);
+                // If we get ABOR handle it directly otherwise hand it over to the executor
+                if (_request.getCommand().equalsIgnoreCase("ABOR")) {
```
Comment (k-2-r) [inline src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java:457]: Not sure how it works but for me every user have a thread of BaseFtpConnection. Inside this thread I dont think a user can a multiple commands execution at the same time. So maybe you just have to remove the pool and simplify your code.

File: src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java (hunk 1 above)
Comment (mvangoor) [reply src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java:335]: The FTP protocol demands that an "ABOR" command can be given at any time and needs to be handled immediately.
The executor is to ensure the I/O thread is ready to handle whatever.

The logic of the code was to ensure single threaded ness by dropping/discarding messages, which is a clear bad idea at this stage (it seems).

Currently the easiest solution to ensure commands are handled in the order they came in (excluding ABOR) is to use a FIFO queue and a single threaded executor.

I agree it is less than ideal, but at least improves the way drftpd handles the control connection.

File: src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java (hunk 2 above)
Comment (mvangoor) [reply src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java:457]: I agree, but so no way to actually handle the "ABOR" use case per above without doing it this way. I am open to suggestions tho

File: src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java (hunk 1 above)
Comment (richard-julien) [reply src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java:335]: Ok thanks for the explanation.
Looks like a good trade off with the current code base, good job.

Comment (k-2-r) [reply src/core/master/src/main/java/org/drftpd/master/network/BaseFtpConnection.java:335]: Ok thanks for the explanation.
Looks like a good trade off with the current code base, good job.

**Pull request merged**